### PR TITLE
Make Person immutable #751

### LINKED ISSUE #751: Make Person immutable
Our code base doesn't use any of Person's mutation methods/features, and will be unlikely to do so in the future due to how easy it is to break invariants by mutating an existing object which we do not own.

Making Person immutable will simplify the class hierarchy (no more ReadOnlyPerson) and remove some code we'll never use. Immutability means we won't have to do defensive copying any more, so we'll be kinder to the GC.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #751, #653, #651.

```
Person is mutable.

This is bad as:
1. It is easy to break invariants by mutating an existing object which
we do not own.
2. No code uses Person’s setter methods.

By making Person immutable, we can make the following improvements:
1. Simplify the class hierarchy as we no longer require ReadOnlyPerson.
2. Remove the need for defensive copying of Person objects.

Let’s make Person immutable.
```

### CODE REVIEW

File: src/test/java/seedu/address/testutil/PersonBuilder.java [hunk 1]
```diff
@@ -42,15 +45,19 @@ public PersonBuilder() {
      * Initializes the PersonBuilder with the data of {@code personToCopy}.
      */
     public PersonBuilder(ReadOnlyPerson personToCopy) {
-        this.person = new Person(personToCopy);
+        name = personToCopy.getName();
+        phone = personToCopy.getPhone();
+        email = personToCopy.getEmail();
+        address = personToCopy.getAddress();
+        tags = personToCopy.getTags();
```
Comment (pyokagan) [inline src/test/java/seedu/address/testutil/PersonBuilder.java]: I think we need to do a defensive copy here. `Person#getTags()` returns an unmodifiable set. https://github.com/pyokagan/addressbook-level4/blob/895430f4c9434c980cf6ed148e0ad36e37c5c5c2/src/main/java/seedu/address/model/person/Person.java#L111

From looking at the rest of `PersonBuilder`, it seems the set will not be modified, but it's better to make ownership clear.

File: src/main/java/seedu/address/model/person/Person.java [hunk 2]
```diff
@@ -18,12 +17,12 @@
...
 public class Person implements ReadOnlyPerson {
 
-    private ObjectProperty<Name> name;
-    private ObjectProperty<Phone> phone;
-    private ObjectProperty<Email> email;
-    private ObjectProperty<Address> address;
+    private final ObjectProperty<Name> name;
+    private final ObjectProperty<Phone> phone;
+    private final ObjectProperty<Email> email;
+    private final ObjectProperty<Address> address;
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/Person.java]: As long as we are using `ObjectProperty` internally, `Person` is still technically mutable.

This should just be plain `final Name`, `final Phone` etc.

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 3]
```diff
@@ -91,7 +90,7 @@ public CommandResult executeUndoableCommand() throws CommandException {
      * Creates and returns a {@code Person} with the details of {@code personToEdit}
      * edited with {@code editPersonDescriptor}.
      */
-    private static Person createEditedPerson(ReadOnlyPerson personToEdit,
+    private static Person createEditedPerson(Person personToEdit,
```
Comment (pyokagan) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Maybe this line could be re-wrapped.

File: src/main/java/seedu/address/model/person/Person.java [hunk 4]
```diff
@@ -38,49 +38,41 @@ public Person(Name name, Phone phone, Email email, Address address, Set<Tag> tag
...
+     * Creates a copy of the given Person.
      */
-    public Person(ReadOnlyPerson source) {
+    public Person(Person source) {
         this(source.getName(), source.getPhone(), source.getEmail(), source.getAddress(),
                 source.getTags());
     }
 
-    @Override
     public ObjectProperty<Name> nameProperty() {
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/Person.java]: There's no point exposing `nameProperty()`, `phoneProperty()` etc. because Person is immutable.

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 5]
```diff
@@ -25,12 +25,12 @@
 
     private final ObservableList<Person> internalList = FXCollections.observableArrayList();
     // used by asObservableList()
-    private final ObservableList<ReadOnlyPerson> mappedList = EasyBind.map(internalList, (person) -> person);
+    private final ObservableList<Person> mappedList = EasyBind.map(internalList, (person) -> person);
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: This may not be required already.

**Comment by pyokagan** [comment]
Developer guide model diagrams also need to be updated in [5/5]

**Comment by pyokagan** [comment]
and all the references to `ReadOnlyPerson` in the developer guide as well.

**Comment by Zhiyuan-Amos** [comment]
Thanks for such a thorough review! :)

> [4/5] The copy constructor of Person can also be removed. Since it is
immutable, there is no point making copies.

I'll do it in [5/5], is that ok? That's because, without the copy constructor in [4/4], I cannot convert a `ReadOnlyPerson` to `Person` easily (many classes are still taking in a `ReadOnlyPerson`, but requires a `Person` for processing some stuffs), and I can only do so by doing `new Person(rop.getName(), rop.getPhone()...)`, which is kinda tedious, especially since I will be removing the construction entirely afterwards.

> Was this done with an automated tool?

There are 4 automated steps required for the entire change, not sure if this is what you're looking for. ><

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java [hunk 6]
```diff
@@ -36,7 +37,7 @@ public EditPersonDescriptorBuilder(ReadOnlyPerson person) {
         descriptor.setPhone(person.getPhone());
         descriptor.setEmail(person.getEmail());
         descriptor.setAddress(person.getAddress());
-        descriptor.setTags(person.getTags());
+        descriptor.setTags(new HashSet<>(person.getTags()));
```
Comment (pyokagan) [inline src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java]: Hmph, this seems like the wrong layer to be doing defensive copying. Shouldn't `EditPersonDescriptor#setTags` be doing the copying?

File: src/test/java/seedu/address/testutil/PersonBuilder.java (hunk 1 above)
Comment (pyokagan) [reply src/test/java/seedu/address/testutil/PersonBuilder.java]: https://github.com/se-edu/addressbook-level4/pull/765#discussion_r158577449

File: src/main/java/seedu/address/model/person/Person.java [hunk 7]
```diff
@@ -1,41 +1,37 @@
...
 
-import javafx.beans.property.ObjectProperty;
-import javafx.beans.property.SimpleObjectProperty;
 import seedu.address.model.tag.Tag;
 import seedu.address.model.tag.UniqueTagList;
 
 /**
  * Represents a Person in the address book.
- * Guarantees: details are present and not null, field values are validated.
+ * Guarantees: details are present and not null, field values are validated, immutability.
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/Person.java]: `Name`, `Phone`, `Address` etc. all use "immutable", so the same should be followed here as well.

File: src/main/java/seedu/address/model/person/Person.java [hunk 8]
```diff
@@ -1,41 +1,37 @@
...
     /**
      * Every field must be present and not null.
      */
     public Person(Name name, Phone phone, Email email, Address address, Set<Tag> tags) {
         requireAllNonNull(name, phone, email, address, tags);
-        this.name = new SimpleObjectProperty<>(name);
-        this.phone = new SimpleObjectProperty<>(phone);
-        this.email = new SimpleObjectProperty<>(email);
-        this.address = new SimpleObjectProperty<>(address);
-        // protect internal tags from changes in the arg list
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/Person.java]: This comment originally came from https://github.com/se-edu/addressbook-level4/commit/bf265bd321dce10ad8d2cfc0203aca8f0d4271c3 (aka the original AB4) so it's better to not remove it.

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 9]
```diff
@@ -44,26 +43,17 @@ public PersonCard(ReadOnlyPerson person, int displayedIndex) {
...
         id.setText(displayedIndex + ". ");
-        initTags(person);
-        bindListeners(person);
+        initPersonDetails(person);
     }
 
     /**
-     * Binds the individual UI elements to observe their respective {@code Person} properties
-     * so that they will be notified of any changes.
+     * Initialises the {@code Label}s with details of {@code person}.
```
Comment (pyokagan) [inline src/main/java/seedu/address/ui/PersonCard.java]: `Initializes`

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 10]
```diff
@@ -44,26 +43,17 @@ public PersonCard(ReadOnlyPerson person, int displayedIndex) {
...
-        address.textProperty().bind(Bindings.convert(person.addressProperty()));
-        email.textProperty().bind(Bindings.convert(person.emailProperty()));
-        person.tagProperty().addListener((observable, oldValue, newValue) -> {
-            tags.getChildren().clear();
-            person.getTags().forEach(tag -> tags.getChildren().add(new Label(tag.tagName)));
-        });
-    }
-
-    private void initTags(ReadOnlyPerson person) {
+    private void initPersonDetails(ReadOnlyPerson person) {
```
Comment (pyokagan) [inline src/main/java/seedu/address/ui/PersonCard.java]: Can just inline the whole method into the constructor. Putting the code in a separate method doesn't add value since it does not introduce a new layer of abstraction.

File: src/main/java/seedu/address/model/AddressBook.java [hunk 11]
```diff
@@ -95,32 +94,32 @@ public void addPerson(ReadOnlyPerson p) throws DuplicatePersonException {
...
+     * @see #syncMasterTagListWith(Person)
      */
-    public void updatePerson(ReadOnlyPerson target, ReadOnlyPerson editedReadOnlyPerson)
+    public void updatePerson(Person target, Person editedPerson)
             throws DuplicatePersonException, PersonNotFoundException {
-        requireNonNull(editedReadOnlyPerson);
+        requireNonNull(editedPerson);
 
-        Person editedPerson = syncMasterTagListWith(editedReadOnlyPerson);
+        Person syncedEditedPerson = syncMasterTagListWith(editedPerson);
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/AddressBook.java]: Sneaky change here, should be mentioned in the commit message.

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 12]
```diff
@@ -24,13 +22,11 @@
 public class UniquePersonList implements Iterable<Person> {
 
     private final ObservableList<Person> internalList = FXCollections.observableArrayList();
-    // used by asObservableList()
-    private final ObservableList<ReadOnlyPerson> mappedList = EasyBind.map(internalList, (person) -> person);
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: This can be in its own commit.

File: src/main/java/seedu/address/model/person/Person.java [hunk 13]
```diff
@@ -66,16 +54,25 @@ public Address getAddress() {
...
+
+        if (!(other instanceof Person)) {
+            return false;
+        }
+
+        Person otherPerson = (Person) other;
+        return otherPerson.getName().equals(this.getName())
+                && otherPerson.getPhone().equals(this.getPhone())
+                && otherPerson.getEmail().equals(this.getEmail())
+                && otherPerson.getAddress().equals(this.getAddress());
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/Person.java]: The movement of `getAsText()` and `isSameStateAs()` from `ReadOnlyPerson` to `Person` should be in its own commit.

File: src/main/java/seedu/address/model/person/Person.java [hunk 14]
```diff
@@ -34,30 +34,18 @@ public Person(Name name, Phone phone, Email email, Address address, Set<Tag> tag
         this.tags = new UniqueTagList(tags);
     }
 
-    /**
-     * Creates a copy of the given ReadOnlyPerson.
-     */
-    public Person(ReadOnlyPerson source) {
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/person/Person.java]: Removal of copy constructor should be in its own commit.

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java (hunk 6 above)
Comment (Zhiyuan-Amos) [reply src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java]: Ah yes, I mixed this up with your other comment https://github.com/se-edu/addressbook-level4/pull/765#discussion_r158577449. I.e. this change shouldn't be made here. 

On a side note, `EditPersonDescriptorBuilder` & `EditPersonDescriptor` have potential bugs since there's no defensive copying.

**Comment by Zhiyuan-Amos** [comment]
Thanks for your review @pyokagan , sorry for missing out some stuffs in the previous iteration ><

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 15]
```diff
@@ -157,7 +157,7 @@ public ReadOnlyAddressBook getAddressBook() {
 
         @Override
         public void addPerson(Person person) throws DuplicatePersonException {
-            personsAdded.add(new Person(person));
+            personsAdded.add(person);
```
Comment (pyokagan) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java]: There should be a non-null assertion here as well to catch errors as soon as possible in tests.

**Comment by Zhiyuan-Amos** [comment]
> No further comments other than the observation that the name syncMasterTagListWith doesn't seem to reflect its true purpose. It sounds like a one-way sync but actually is a two-way sync?

Yup hmm, does `syncWithMasterTagList` sound better?

**Comment by damithc** [comment]
> Yup hmm, does syncWithMasterTagList sound better?

Yes, that's better.

**Comment by Zhiyuan-Amos** [comment]
@damithc for your review in case you forgot, since you reviewed my other PR but not this one :P
